pkg/util/logger/logger.go

```diff
@@ -10,35 +14,95 @@ import (
+	debugLogLevelElevation = 1
+)
 
 var (
-	baseLogger     logr.Logger
+	baseLogger     DtLogger
 	baseLoggerOnce sync.Once
 )
 
 // Get returns a new, unnamed logger configured with the basics we need for operator logs which can be used as a blueprint for
 // derived loggers int the operator components.
-func Get() logr.Logger {
+func Get() DtLogger {
 	baseLoggerOnce.Do(func() {
-		// its important to create only one "main" logger to avoid excessive memory usage, creating a full logger is rather expensive,
-		// deriving other loggers by WithName is rather cheap
-		config := zap.NewProductionEncoderConfig()
-		config.EncodeTime = zapcore.ISO8601TimeEncoder
-		config.StacktraceKey = stacktraceKey
-		baseLogger = ctrlzap.New(ctrlzap.WriteTo(NewPrettyLogWriter()), ctrlzap.Encoder(zapcore.NewJSONEncoder(config)), ctrlzap.Level(readLogLevelFromEnv()))
+		logLevel := readLogLevelFromEnv()
+		baseLogger = createLogger(NewPrettyLogWriter(), logLevel)
+		baseLogger.Info("logging level", "logLevel", logLevel.String())
 	})
 
 	return baseLogger
 }
 
+func createLogger(out io.Writer, logLevel zapcore.Level) DtLogger {
+	// its important to create only one "main" logger to avoid excessive memory usage, creating a full logger is rather expensive,
+	// deriving other loggers by WithName is rather cheap
+	config := zap.NewProductionEncoderConfig()
+	config.EncodeTime = zapcore.ISO8601TimeEncoder
+	config.StacktraceKey = stacktraceKey
+
+	return DtLogger{
+		ctrlzap.New(ctrlzap.WriteTo(out), ctrlzap.Encoder(zapcore.NewJSONEncoder(config)), ctrlzap.Level(logLevel)),
+	}
+}
+
 func readLogLevelFromEnv() zapcore.Level {
 	envLevel := os.Getenv(LogLevelEnv)
 
 	level, err := zapcore.ParseLevel(envLevel)
-	if err != nil {
+	if err != nil || len(envLevel) == 0 {
 		level = zapcore.DebugLevel
 	}
 
 	return level
 }
+
+type DtLogger struct {
+	logr.Logger
+}
+
+// Debug can be used for verbose output that is supposed to be  valuable for troubleshooting
+func (l *DtLogger) Debug(message string, keysAndValues ...any) {
+	kv := make([]any, 0)
+	kv = append(kv, keysAndValues...)
+	kv = append(kv, "caller", getCaller())
+
+	l.debugLog(message, kv...)
+}
+
+func (l DtLogger) WithName(name string) DtLogger {
+	return DtLogger{l.Logger.WithName(name)}
+}
+
+func (l DtLogger) WithValues(keysAndValues ...any) DtLogger {
+	return DtLogger{l.Logger.WithValues(keysAndValues...)}
+}
+
+func getCaller() string {
+	const callerNameOffset = 2
+
+	if pc, _, _, ok := runtime.Caller(callerNameOffset); ok {
+		details := runtime.FuncForPC(pc)
+		filePath, line := details.FileLine(pc)
+		fileName := filepath.Base(filePath)
+		functionName := filepath.Base(details.Name())
+
+		return fmt.Sprintf("%s (%s:%d)", functionName, fileName, line)
+	}
+
+	return "<unknown function>"
+}
+
+func (l *DtLogger) DebugLogFunctionBoundaries(keysAndValues ...any) func() {
+	l.debugLog("enter "+getCaller(), keysAndValues...)
+
+	return func() {
+		l.debugLog("leave "+getCaller(), keysAndValues...)
+	}
+}
```

Comment:
not the biggest fan of this, as it just adds a stacktrace to the debug log, when its not necessarily needed.

which is worsened by the fact that often its used with other debug logs that also signal the start/end of a function (I commented about one previously)

Reply:
I've added this intentionally to this PR to have something to discuss on. And honestly also to me it feels a bit like constant printf-debugging. It might be useful to have this at hand during development, but IMHO it should be used very sparsely in prod.
In contrary to a classic stack trace, it provides some kind of bracket (aka. context) to other log lines in between and it's also available when there's no actual error thrown.

Reply:
I don't see any benefits of it when you have debugger instead 🤔

Reply:
I removed this function from this PR if we really need it we can decide later.